File: iou-cordapp/src/main/java/com/atfx/sample/flow/FinalizeFlow.java
package com.atfx.sample.flow;

import co.paralleluniverse.fibers.Suspendable;
import com.atfx.sample.utils.CordaUtils;
import net.corda.core.contracts.ContractState;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.contracts.StateRef;
import net.corda.core.flows.*;
import net.corda.core.identity.AbstractParty;
import net.corda.core.identity.Party;
import net.corda.core.transactions.SignedTransaction;
import net.corda.core.utilities.ProgressTracker;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.stream.Collectors;

@InitiatingFlow
public class FinalizeFlow extends FlowLogic<SignedTransaction> {

    private SignedTransaction stx;
    ProgressTracker progressTracker;

    public FinalizeFlow(SignedTransaction stx) {
        this(stx, FinalityFlow.tracker());
    }

    public FinalizeFlow(SignedTransaction stx,
                        ProgressTracker progressTracker) {
        super();
        this.stx = stx;
        this.progressTracker = progressTracker;
    }

    @Override
    @Suspendable
    public SignedTransaction call() throws FlowException {
        List<StateAndRef<ContractState>> inputStates = new ArrayList<>();
        for (StateRef inputState : stx.getInputs()) {
            inputStates.add(getServiceHub().toStateAndRef(inputState));
        }
        List<AbstractParty> inParticipants = inputStates.stream()
                .flatMap(it -> it.getState().getData().getParticipants().stream())
                .collect(Collectors.toList());

        List<AbstractParty> outParticipants =
                stx.getTx().outRefsOfType(ContractState.class).stream()
                        .flatMap(it -> it.getState().getData().getParticipants().stream())
                        .collect(Collectors.toList());

        ArrayList<AbstractParty> allParticipants = new ArrayList<>(inParticipants);
        allParticipants.addAll(outParticipants);

        HashSet<Party> allParty = new HashSet<>();
        for(AbstractParty ap: allParticipants) {
            Party party = getServiceHub().getIdentityService()
                    .requireWellKnownPartyFromAnonymous(ap);
            allParty.add(party);
        }
        allParty.remove(CordaUtils.getServiceHubMyIdentity(getServiceHub()));

        ArrayList<FlowSession> sessions = new ArrayList<>();
        for (Party party: allParty) {
            sessions.add(initiateFlow(party));
        }
        return subFlow(new FinalityFlow(stx, sessions));
    }

    public ProgressTracker tracker() {
        return FinalityFlow.tracker();
    }
}

File: iou-cordapp/src/main/kotlin/com/formssihk/sample/flow/FinalizeFlow.kt
package com.formssihk.sample.flow

import co.paralleluniverse.fibers.Suspendable
import net.corda.core.contracts.ContractState
import net.corda.core.flows.*
import net.corda.core.node.ServiceHub
import net.corda.core.transactions.SignedTransaction
import net.corda.core.transactions.TransactionBuilder
import net.corda.core.utilities.ProgressTracker

@InitiatingFlow
class FinalizeFlow(
    private val stx: SignedTransaction,
    progressTracker: ProgressTracker = FinalityFlow.tracker()
): FlowLogic<SignedTransaction>() {
    companion object {
        fun tracker() = FinalityFlow.tracker()
    }

    @Suspendable
    override fun call(): SignedTransaction {
        val inParticipants = stx.inputs.map { serviceHub.toStateAndRef<ContractState>(it) }.flatMap { it.state.data.participants }
        val outParticipants = stx.tx.outRefsOfType<ContractState>().flatMap { it.state.data.participants }
        val allParticipants = inParticipants + outParticipants
        val sessions = allParticipants.map { serviceHub.identityService.requireWellKnownPartyFromAnonymous(it) }.distinct()
            .filterNot { it == serviceHub.myInfo.legalIdentities.first() }
            .map { initiateFlow(it) }
        return subFlow(FinalityFlow(stx, sessions))
    }

}

@InitiatedBy(FinalizeFlow::class)
class ReceiveFinalizeFlow(val otherSide: FlowSession) : FlowLogic<SignedTransaction>() {
    @Suspendable
    override fun call() = subFlow(ReceiveFinalityFlow(otherSide))
}

fun TransactionBuilder.fillContractAttachments(serviceHub: ServiceHub) {
    /*
    listOf(
        // serviceHub.attachments.internalFindTrustedAttachmentForClass("com/r3/demos/ubin2a/base/WalletType")?.id,
        //serviceHub.attachments.getLatestContractAttachments(Cash.PROGRAM_ID).firstOrNull(),
        //serviceHub.attachments.getLatestContractAttachments(Instruction.PROGRAM_ID).firstOrNull(),
        //serviceHub.attachments.getLatestContractAttachments("th.or.bot.inthanon.bond.bond.contracts.Bond").firstOrNull(),
        //serviceHub.attachments.getLatestContractAttachments("th.or.bot.inthanon.mlsm.contract.ProposeRequest").firstOrNull(),
        //serviceHub.attachments.getLatestContractAttachments("th.or.bot.inthanon.scheduler.contracts.SchedulerJob").firstOrNull(),
        //serviceHub.attachments.getLatestContractAttachments("th.or.bot.inthanon.remittance.contracts.Remittance").firstOrNull()
    ).filter { it != null }.forEach {
        this.addAttachment(it!!)
    }*/
}